cartographer/mapping/map_builder.cc

```diff
@@ -350,24 +362,45 @@ void MapBuilder::LoadMap(io::ProtoStreamReaderInterface* const reader) {
           trajectory_remapping.at(proto.trajectory_data().trajectory_id()));
       pose_graph_->SetTrajectoryDataFromProto(proto.trajectory_data());
     }
-    // TODO(ojura): Deserialize IMU, odometry and fixed frame pose data when
-    // loading unfrozen trajectories.
+    if (!load_frozen_state) {
+      if (proto.has_imu_data()) {
+        pose_graph_->AddImuData(
+            trajectory_remapping.at(proto.imu_data().trajectory_id()),
+            sensor::FromProto(proto.imu_data().imu_data()));
+      }
+      if (proto.has_odometry_data()) {
+        pose_graph_->AddOdometryData(
+            trajectory_remapping.at(proto.odometry_data().trajectory_id()),
+            sensor::FromProto(proto.odometry_data().odometry_data()));
+      }
+      if (proto.has_fixed_frame_pose_data()) {
+        pose_graph_->AddFixedFramePoseData(
+            trajectory_remapping.at(
+                proto.fixed_frame_pose_data().trajectory_id()),
+            sensor::FromProto(
+                proto.fixed_frame_pose_data().fixed_frame_pose_data()));
+      }
+      // TODO(pifon2a, ojura): deserialize landmarks
+    }
   }
 
-  // Add information about which nodes belong to which submap.
-  for (const proto::PoseGraph::Constraint& constraint_proto :
-       pose_graph_proto.constraint()) {
-    if (constraint_proto.tag() !=
-        mapping::proto::PoseGraph::Constraint::INTRA_SUBMAP) {
-      continue;
+  if (load_frozen_state) {
+    // Add information about which nodes belong to which submap.
+    for (const proto::PoseGraph::Constraint& constraint_proto :
+         pose_graph_proto.constraint()) {
+      if (constraint_proto.tag() !=
+          mapping::proto::PoseGraph::Constraint::INTRA_SUBMAP) {
+        continue;
+      }
+      pose_graph_->AddNodeToSubmap(
+          NodeId{constraint_proto.node_id().trajectory_id(),
+                 constraint_proto.node_id().node_index()},
+          SubmapId{constraint_proto.submap_id().trajectory_id(),
+                   constraint_proto.submap_id().submap_index()});
     }
-    const NodeId node_id{
-        trajectory_remapping.at(constraint_proto.node_id().trajectory_id()),
-        constraint_proto.node_id().node_index()};
-    const SubmapId submap_id{
-        trajectory_remapping.at(constraint_proto.submap_id().trajectory_id()),
-        constraint_proto.submap_id().submap_index()};
-    pose_graph_->AddNodeToSubmap(node_id, submap_id);
+  } else {
```

Comment:
@cschuet Maybe we can add a comment here that `AddSerializedConstraints` performs all the work in the branch above (when `load_frozen_state` is true), and more (actually adds all the constraints). If I remember correctly, node-to-submap information is required for 3D pure lozalization. Restoring all constraints would restore this information, but it would be wasteful, so @wohe added `AddNodeToSubmap` which adds only the node-to-submap information.

When we are doing full unfrozen state restoration, we are actually restoring all constraints, which results in adding the node-to-submap information as well, so we don't have to bother with `AddNodeToSubmap`.